### DISCUSSION TITLE
Call server functions with the manifest hash

### DISCUSSION THREAD

**Discussion by wmertens** [QwikDev/qwik-evolution#19] [Proposals For Qwik]
### What is it about?

Prevent version skew in deployments

#### Problems you are trying to solve:
* Mismatched client versions calling server functions
* Allow A/B testing

#### Goals you are trying to achieve:
* Ensure that no calls can happen with potentially incompatible data/handlers
* Ability to route clients to their server version

#### Any other context or information you want to share:
I believe Vercel or Next can do skew control based on headers.

---

#### What do you propose?
Every call to the server should include a token that encodes the client version/deployment. It should default to the manifest hash.

This can then be used on the server side to decline calls from different tokens.

This doubles as a mechanism to always route clients to the same deployment.

0. SSR puts token on container, defaulting to the manifest hash
1. client calls server action and passes the token in a header
2. (optional) the service reads the token and chooses a deployment
3. the deployment verifies the token and refuses the the call if it doesn't match
4. when the call is refused, the client reloads the page, allowing the user to try again.

Note, 4. should happen as little as possible. Devs should ensure that clients can keep talking to compatible servers while notifying them they should reload, or silently reloading when the time is right.

Passing your own token also allows overriding the mechanism, so that a single server can handle multiple versions. This obviously means that the dev is responsible for version mismatches in action handling.

### Links / References

https://vercel.com/blog/version-skew-protection

**Comment by jakovljevic-mladen** [comment]
This is great proposal. I've noticed this issue with qwik.dev site where I have it open in one tab for couple of days and then when I go to that tab after some time and want to use search, it simply doesn't work, probably because it was skewed. Then, I have to manually refresh page for it to work.

However, I don't know how Qwik can work on this, but step 3

> the deployment verifies the token and refuses the the call if it doesn't match

instead of refusing, should be able to find server instance and use it so that users don't notice anything. This is largely dependent on the infrastructure and developers should make sure they have multiple server instances running and they should probably have load balancer that would decide where to redirect incoming request.

Have you considered doing something like this, or at least providing a way for users to use version hash so they can build their infrastructure around it?

**Reply by wmertens** [reply]
The redirect should happen in a proxy before it hits the server. Qwik-city doesn't have an opinion of how to deploy.
